src/libraries/Native/Unix/System.Native/pal_memory.c

```diff
@@ -6,22 +6,45 @@
 #include <stdlib.h>
 #include <string.h>
 
-void* SystemNative_MemAlloc(uintptr_t size)
+void* SystemNative_AlignedAlloc(uintptr_t alignment, uintptr_t size)
 {
-    return malloc(size);
+#if HAVE_ALIGNED_ALLOC
+    return aligned_alloc(alignment, size);
+#elif
+    void* result = nullptr;
+    posix_memalign(&result, alignment, size);
+    return result;
+#else
+    #error "Platform doesn't support aligned_alloc or posix_memalign"
```

Comment:
I have it erroring if `posix_memalign` isn't available as a fallback. `posix_memalign` is a very old (`_POSIX_C_SOURCE >= 200112L || _XOPEN_SOURCE >= 600`) function and should be always available.

Likewise, I don't believe trying to polyfill via `malloc`/`free` is reliable as there is no guarantee that `free` works with arbitrary pointers:
> The behavior is undefined if the value of ptr does not equal a value returned earlier by malloc(), calloc(), realloc(), or aligned_alloc() (since C11).

Likewise, there is no reliable way to backtrack from a given aligned `ptr` to the actual `ptr` that `malloc` did return.

Reply:
>  I don't believe trying to polyfill via malloc/free is reliable as there is no guarantee that free works with arbitrary pointers:

The polyfill would have to be for both aligned alloc and free, something like this: https://github.com/dotnet/runtime/issues/33244#issuecomment-595848832

Reply:
I'd like to see if any platforms don't actually have `posix_memalign` before falling back to storing the actual pointer before the returned pointer.